lib/device-sockets/ios/socket-proxy-factory.ts

```diff
@@ -0,0 +1,135 @@
+import * as net from "net";
+import semver = require("semver");
+import temp = require("temp");
+import ws = require("ws");
+
+export class SocketProxyFactory implements ISocketProxyFactory {
+	constructor(private $logger: ILogger,
+		private $projectData: IProjectData,
+		private $projectDataService: IProjectDataService) { }
+
+	public createSocketProxy(factory: () => net.Socket, socketFactoryAction: (backendSocket: net.Socket, frontendSocket: net.Socket) => void): IFuture<any> {
+		return (() => {
+			let socketFactory = (callback: (_socket: net.Socket) => void) => this.connectEventually(factory, callback);
+
+			this.$projectDataService.initialize(this.$projectData.projectDir);
+			let frameworkVersion = this.$projectDataService.getValue("tns-ios").wait().version;
+			let result: any;
+
+			if(semver.gte(frameworkVersion, "1.4.0")) {
+				result = this.createTcpSocketProxy(socketFactory, socketFactoryAction);
+			} else {
+				result = this.createWebSocketProxy(socketFactory);
+			}
+
+			return result;
+		}).future<any>()();
+	}
+
+	private connectEventually(factory: () => net.Socket, handler: (_socket: net.Socket) => void) { // TODO: Add socketProxyBase
+		function tryConnect() {
+			let tryConnectAfterTimeout = setTimeout.bind(undefined, tryConnect, 1000);
+
+			let socket = factory();
+			socket.on("connect", () => {
+				socket.removeListener("error", tryConnectAfterTimeout);
+				handler(socket);
+			});
+			socket.on("error", tryConnectAfterTimeout);
+		}
+
+		tryConnect();
+	}
+
+	private createWebSocketProxy(socketFactory: (handler: (socket: net.Socket) => void) => void): ws.Server {
+		// NOTE: We will try to provide command line options to select ports, at least on the localhost.
+		let localPort = 8080;
+
+		this.$logger.info("\nSetting up debugger proxy...\nPress Ctrl + C to terminate, or disconnect.\n");
+
+		// NB: When the inspector frontend connects we might not have connected to the inspector backend yet.
+		// That's why we use the verifyClient callback of the websocket server to stall the upgrade request until we connect.
+		// We store the socket that connects us to the device in the upgrade request object itself and later on retrieve it
+		// in the connection callback.
+
+		let server = ws.createServer(<any>{
+			port: localPort,
+			verifyClient: (info: any, callback: any) => {
+				this.$logger.info("Frontend client connected.");
+				socketFactory((_socket: any) => {
+					this.$logger.info("Backend socket created.");
+					info.req["__deviceSocket"] = _socket;
+					callback(true);
+				});
+			}
+		});
+		server.on("connection", (webSocket) => {
+			let deviceSocket: net.Socket = (<any>webSocket.upgradeReq)["__deviceSocket"];
+			let packets = new PacketStream();
+			deviceSocket.pipe(packets);
+
+			packets.on("data", (buffer: Buffer) => {
+				webSocket.send(buffer.toString("utf16le"));
+			});
+
+			webSocket.on("message", (message, flags) => {
+				let length = Buffer.byteLength(message, "utf16le");
```

Comment:
Can we document this code, there's no clue why it fills the buffer with these values.